upgrade-2.x.sh

```diff
@@ -0,0 +1,201 @@
+#!/bin/bash
+
+set -o errexit
+set -o pipefail
+
+TARGET_VERSION=$1
+PREFERRED_HOSTOS_VERSION=2.0.7
+
+. /etc/os-release
+
+DOCKER=docker
+
+# Log function helper
+function log {
+    # Address log levels
+    case $1 in
+        ERROR)
+            loglevel=ERROR
+            shift
+            ;;
+        WARN)
+            loglevel=WARNING
+            shift
+            ;;
+        *)
+            loglevel=LOG
+            ;;
+    esac
+    ENDTIME=$(date +%s)
+    if [ "z$LOG" == "zyes" ] && [ -n "$LOGFILE" ]; then
+        printf "[%09d%s%s\n" "$(($ENDTIME - $STARTTIME))" "][$loglevel]" "$1" | tee -a $LOGFILE
+    else
+        printf "[%09d%s%s\n" "$(($ENDTIME - $STARTTIME))" "][$loglevel]" "$1"
+    fi
+    if [ "$loglevel" == "ERROR" ]; then
+        /usr/bin/resin-device-progress --percentage 100 --state "ResinOS: Update failed."
+        exit 1
+    fi
+}
+
+# Test if a version is greater than another
+function version_gt() {
+    test "$(echo "$@" | tr " " "\n" | sort -V | head -n 1)" != "$1"
+}
+
+# Log timer
+STARTTIME=$(date +%s)
```

Comment:
non-exported functions should be lowercase, to indicate they are non-exported and make the code more readable.